### ISSUE TITLE
Rule Proposal: no-invalid-this

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#2815]
Hello.

I thought when I saw https://github.com/babel/babel/issues/1795: we need a rule about `this` keyword.

The `this` keyword is `undefined` by default under the strict mode.
So if `this` is `undefined` clearly, this rule warns the `this`.

But the judgement is very difficult... (cannot?)
To start, I suggest independent functions are warned.

## The following patterns are considered warnings:

- A
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  console.log(this);
  bar(() => console.log(this));
  ```

- B
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  function foo() {
      console.log(this);
      bar(() => console.log(this));
  }
  ```

- C
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  var foo = function() {
      console.log(this);
      bar(() => console.log(this));
  };
  ```

- D
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  foo(function() {
      console.log(this);
      bar(() => console.log(this));
  });
  ```

- E
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  var obj = {
      foo: () => {
          console.log(this);
          bar(() => console.log(this));
      }
  };
  ```

- F
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  obj.foo = () => {
      console.log(this);
      bar(() => console.log(this));
  };
  ```

## The following patterns are not considered warnings:

- G
  ```js
  // {ecmaFeatures: {modules: false}}
  console.log(this);
  bar(() => console.log(this));
  ```

- H
  ```js
  // {ecmaFeatures: {modules: false}}
  function foo() {
      console.log(this);
      bar(() => console.log(this));
  }
  ```

- I
  ```js
  // {ecmaFeatures: {modules: false}}
  var foo = function() {
      console.log(this);
      bar(() => console.log(this));
  };
  ```

- J
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  var obj = {
      foo: function foo() {
          console.log(this);
          bar(() => console.log(this));
      },
      bar() {
          console.log(this);
          bar(() => console.log(this));
      }
  };
  ```

- K
  ```js
  // {ecmaFeatures: {modules: false}}
  obj.foo = function() {
      console.log(this);
      bar(() => console.log(this));
  };
  ```

- L
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  class A {
      foo() {
          console.log(this);
          bar(() => console.log(this));
      }
  }
  ```

- M
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  class A {
      foo() {
          var obj = {
              foo: () => {
                  console.log(this);
                  bar(() => console.log(this));
              }
          };
      }
  }
  ```

- N
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  class A {
      foo() {
          obj.foo = () => {
              console.log(this);
              bar(() => console.log(this));
          };
      }
  }
  ```

- O
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  var foo = function() {
      console.log(this);
      bar(() => console.log(this));
  }.bind(x);
  ```

- P
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  foo(function() {
      console.log(this);
      bar(() => console.log(this));
  }.bind(x));
  ```

I want to not warn below, but it's exceptions of rule, so I'm not sure it's good.

- Q
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  xs.forEach(function() {
      console.log(this); // this is y.
      bar(() => console.log(this));
  }, y);
  ```

- R
  ```js
  "use strict"; // or {ecmaFeatures: {modules: true}}
  dom.addEventListener(function() {
      console.log(this); // this is dom.
      bar(() => console.log(this));
  });
  ```

I'm happy to work for this.

Thank you.

**Comment by mathieumg** [comment]
If often use arrow functions and refer to the `this` of the "parent scope" on purpose, and sometimes with a few nesting levels. I don't see how you would be able to determine the intention from static analysis here.

**Comment by michaelficarra** [comment]
@mysticatea I think you're confused about how modules affect `this`. You're right, `this` in top-level arrows in strict mode will always be an `undefined` value, and will be a ReferenceError in modules. But the function declarations and function expressions can be `call`/`apply`'d or stuck on some object and called as `a.b()`. This *would* give them a meaningful `this` value. I think you should reduce this proposal to a a much simpler one that's just about top-level `this` and arrows that mention `this`, and then we can see if we'd like to expand it further in the future (which I doubt we'd want to do except behind an option).

**Comment by mysticatea** [comment]
* What determines `this` is NOT "there is top-level or not".
  It's "the direct containing function (without arrow functions) of `this` is called as a method or not".

  For example:

  ```js
  "use strict";

  console.log(this); // undefined. upper scope is not a method.
  
  function foo() {
      console.log(this); // undefined. upper scope is not a method.
  }
  foo();

  var obj = {
      foo: function() {
          function bar() {
              console.log(this); // undefined. upper scope is not a method.
          }
          bar();
      }
  };
  obj.foo();

  class A {
      foo() {
          function bar() {
              console.log(this); // undefined. upper scope is not a method.
          }
          bar();
      }
  }
  var a = new A();
  a.foo();
  ```

* Of course, I know the `call/apply` callings.  But I think those are not generally.  This rule considers the `call/apply` callings are special case.  
  This is similar to TypeScript.  So far as I know, the function type literal of TypeScript cannot specify the type of `this`.

So this rule is simple.
This rule checks whether or not the direct containing function (without arrow functions) of `this` is a method (property value).

**Comment by michaelficarra** [comment]
@mysticatea Firstly, `call`/`apply` are used *all the time*. Secondly, I don't think you're considering the common patterns of `existingObject.a = function(){ this; };` or `Object.defineProperty(o, "a", { writable: false, value: function(){ this; } }` or many others.

**Comment by mysticatea** [comment]
@michaelficarra Ah, thank you, I had forgotten `Object.defineProperty`. `existingObject.a = function(){ this; };` is the pattern K....

I will add valid patterns:

* S
  ```js
  Object.defineProperty(obj, "foo", {value: function() {
      this;
      bar(() => this);
  });
  ```

* T
  ```js
  Object.defineProperties(obj, {
      foo: {value: function() {
          this;
          bar(() => this);
      }}
  });
  ```

* U
  ```js
  obj.foo = (function() {
      return function foo() {
          this;
          bar(() => this);
      };
  }());
  ```

* V
  ```js
  var obj = {
      foo: (function() {
          return function foo() {
              this;
              bar(() => this);
          };
      }())
  };
  ```

And [JSDoc `@this` tag](http://usejsdoc.org/tags-this.html)

* W

  ```js
  /** @this {Foo} */
  function bar() {
    this;
    baz(() => this);
  }
  ```

**Comment by nzakas** [comment]
I'm not sure about this - I think you are trying to catch too many different cases. I agree with @michaelficarra that the scope of this rule needs to be reduced before moving forward.

**Comment by mysticatea** [comment]
I also think this rule looks complex, so yesterday I checked whether or not the implementation of this rule is enough simple.

Before it, I thought I don't need to limit to under strict mode because unexpected accesses to the global scope is dangerous.  In short, I will move the patterns G-I to warned.  This rule doesn't need to check whether or not there is strict mode, so becomes simpler.

And so, I wrote a code: https://gist.github.com/mysticatea/e085a7ef8d67d9e29d6c (without patterns Q, R)

I think this is enough simple.  If I reduce patterns, the body of the `isPropertyValue(node)` function reduces only a few lines.

----

The purpose of this rule is very simple: to avoid unexpected `this` accesses.
But if `this` accesses that I want are warned, it's stressful...
I want to avoid it as accurately as possible.

**Comment by mysticatea** [comment]
Oops, I will add a valid pattern: Constructor.

```js
// the name starts with upper-case.
function Foo() {
  this;
  baz(() => this);
}
```

**Comment by mysticatea** [comment]
I arranged information: https://gist.github.com/mysticatea/2e112dccbfa3a1667995

**Comment by nzakas** [comment]
Looks nice and simple now, let's move forward.

**Comment by mysticatea** [comment]
Thank you!

Can I send a PR for this?

**Comment by michaelficarra** [comment]
I think the rule should be the very simple case of `this` in module context (in a module, but not nested in a FE, FD, GFE, GFD, getter, setter, or method) because it is the only thing we can guarantee is an error. Then we can have options to expand the rule to these other cases that are *likely* errors in common practice, but not necessarily errors.

**Comment by nzakas** [comment]
I agree, good description @michaelficarra 

@mysticatea yes, you can start a PR
